### ISSUE TITLE
4.13.1 (90346) Docker crash when running container with selenium and chrome

### ISSUE THREAD

**Issue by trevesz** [docker/for-mac#6530]
- [X] I have tried with the latest version of Docker Desktop
  - [X] I have tried disabling enabled experimental features
  - [X] I have uploaded Diagnostics
  - Diagnostics ID: 5b5ed26e-0021-4de6-8651-9305b4039bb3/20221021175526

### Expected behavior
Single run python script executes and uses selenium and chrome to scrape target website then done.  Been doing this for almost 3 years daily.

### Actual behavior
After upgrading to Docker Desktop for Mac 4.13.0, the docker engine crashes with this message in the terminal that launched the container with docker run command.  I thought the issue might have been fixed in 4.13.1 but it is still occurring, though less frequently/reliably than before.

`ERRO[0007] error waiting for container: invalid character 'c' looking for beginning of value`

The menu bar icon shows the engine stopping...the dashboard shows engine stopped but trying to restart from the docker menu fails with a fatal error.  Screenshot attached.  You have to exit/quit docker and restart it for the engine to become operational again.

<img width="666" alt="Screen Shot 2022-10-21 at 1 59 38 PM" src="https://user-images.githubusercontent.com/477527/197259869-0039259a-5c6d-4381-9d64-db12133a035c.png">

### Information

  - macOS Version: Monterey 12.6
  - Intel chip or Apple chip: Intel
  - Docker Desktop Version: 4.13.0 & 4.13.1

### Output of `/Applications/Docker.app/Contents/MacOS/com.docker.diagnose check`
This output is before running the container and causing the crash.  If I run this after the crash, of course it generates a bunch of errors about not being able to reach the docker services. I can post that as well but figured you cared more to know that before each run, the base docker installation is healthy.

Starting diagnostics

[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0014: are the backend processes running?
[PASS] DD0007: is the backend responding?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[PASS] DD0012: is the VM networking working?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0019: is the com.docker.vmnetd process responding?
[PASS] DD0033: does the host have Internet access?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0032: do Docker networks overlap with host IPs?
No fatal errors detected.

### Steps to reproduce the behavior

  1. Dockerfile

```
# For more information, please refer to https://aka.ms/vscode-docker-python
FROM python:3.10-slim

# Install packages we are going to need for setting up Chrome
RUN apt-get update && apt-get install -y \
    gnupg \
    wget \
    curl \
    unzip \
    && rm -rf /var/lib/apt/lists/*

# Install Google Chrome
RUN wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add -
RUN sh -c 'echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google-chrome.list'
RUN apt-get update && apt-get install -y \
    google-chrome-stable \
    && rm -rf /var/lib/apt/lists/*

# Install chromedriver
RUN wget -O /tmp/chromedriver.zip http://chromedriver.storage.googleapis.com/`curl -sS chromedriver.storage.googleapis.com/LATEST_RELEASE`/chromedriver_linux64.zip
RUN unzip /tmp/chromedriver.zip chromedriver -d /usr/local/bin/ && rm /tmp/chromedriver.zip

# Set display port to avoid crashing Chrome
ENV DISPLAY=:99

# Keeps Python from generating .pyc files in the container
ENV PYTHONDONTWRITEBYTECODE=1

# Turns off buffering for easier container logging
ENV PYTHONUNBUFFERED=1

# Install pip requirements
ADD requirements.txt .
RUN python -m pip install -r requirements.txt

WORKDIR /app
ADD ./src /app

# Switching to a non-root user, please refer to https://aka.ms/vscode-docker-python-user-rights
RUN useradd appuser && chown -R appuser /app
USER appuser

# During debugging, this entry point will be overridden. For more information, please refer to https://aka.ms/vscode-docker-python-debug
CMD ["python", "app.py"]
```
  2. Run this app.py to cause the fault, you may have to run it a few times before the crash will occur but it only takes 1-3 tries reliably on my machine.
```
from selenium import webdriver
from selenium.webdriver.common.by import By
from time import sleep

class Scraper:
    def __init__(self, delay):
        # Let's grab our default settings for Chromium instance
        chrome_options = self.__build_chrome_options()
        self.delay = delay

        try:
            # Let's instantiate our scraper
            self.driver = webdriver.Chrome(options=chrome_options)
        except Exception as e:
            raise e

    def __build_chrome_options(self):
        chrome_options = webdriver.ChromeOptions()
        chrome_options.add_argument('--window-size=1920,1200')
        chrome_options.add_argument('--headless')
        chrome_options.add_argument('--single-process')
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-gpu')
        chrome_options.add_argument('--enable-logging')
        chrome_options.add_argument('--user-data-dir=/tmp/user1')
        chrome_options.add_argument('--v=1')
        chrome_options.add_argument('--ignore-certificate-errors')
        chrome_options.add_argument('--disable-dev-shm-usage')

        return chrome_options

    def scroll_down(self, delay):
        while True:
            # Get scroll height.
            lastHeight = self.driver.execute_script("return document.body.scrollHeight")

            # Scroll down to the bottom. 
            print(f"Last height: [{lastHeight}] - Scrolling...") # uncomment for debugging scrolling
            self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")

            # Wait to load the page
            sleep(delay/2)

            # Calculate new scroll height and compare with last scroll height
            newHeight = self.driver.execute_script("return document.body.scrollHeight")

            # If the browser hasn’t scrolled any more (i.e. it’s reached the end) then stop
            if newHeight == lastHeight:
                print("Reached bottom") # uncomment for debugging scrolling
                break

    def get_with_delay(self, request_url, delay):
        self.driver.get(request_url)
        sleep(delay)
        self.scroll_down(delay)
    
    def __del__(self):
        self.driver.quit()

# use this for testing locally
if __name__ == "__main__":
    scraper = Scraper(3)

    url = "https://abcnews.go.com/"

    try:
        print("Getting url: ", url)
        scraper.get_with_delay(url, 3)
    except Exception as e:
        print(f'Error {e} while scraping url: {url}')
```

You should just need to add selenium to a requirements.txt file and/or install it on running container to get the test code to work on top of the base python image.  I tried to see if the issue was python, base linux in container and/or selenium version rooted but same result with python 3.8, 3.9 and 3.10, selenium 4.2, 4.3, 4.4, 4.5 and both buster and bullseye distros.  Hope this is helpful in fixing for future Docker release.

**Comment by trevesz** [comment]
I should also point out that I tried this with both chrome_options.add_argument('--disable-dev-shm-usage') option, which is how I've run for years, as well as without that option and added --shm-size=2g to docker run as well.  These are related to memory handling issues and crashes in the past with selenium and chrome but issue occurs with either approach in this case.

**Comment by boxshadow** [comment]
same issue

**Comment by handelaar2** [comment]
same issue. 4.13.0 (89412) crashes
`ERRO[0014] error waiting for container: invalid character 'c' looking for beginning of value`

**Comment by E-G-C** [comment]
same here `invalid character 's' looking for beginning of value`

**Comment by henriquebremenkanp** [comment]
I'm also experiencing this problem after the latest update to 4.13.0. Tried factory reset many times, but always get the same error after some time of usage. 
I guess it was after running `docker-compose up --build`, but I'm not sure. Have no idea on how to debug it...

As it was mentioned above, 4.12.0 works fine. I'm also on Macbook Pro M1 Pro.

**Comment by jyoti-c** [comment]
Same issue with Monterey 12.6.1.

**Comment by oliveremberton** [comment]
Same issue here with Monterey 12.6.1, confirmed for three separate developers.

**Comment by StefanIliev545** [comment]
Same issue M1 Pro, Ventura

**Comment by hopeseekr** [comment]
I can confirm experiencing the same issue with custom built docker image containing both Chromium and Selenium.

**Comment by riabiy-denis** [comment]
> I can confirm experiencing the same issue with custom built docker image containing both Chromium and Selenium. Version 4.13.0 (89412) on Mac OSX (Intel and M1).
> 
> Everything works fine in Docker Desktop v4.12.0.
> 
> We experience it even after rebuilding the images, fwiw.

Thank you for giving an advice on what to downgrade :D 
For some reason I was playing with images, but seems like it's bout new docker release.

**Comment by apotrebin** [comment]
Seems like fixed on Version 4.13.1 
`ERRO[0019] error waiting for container: invalid character 'c' looking for beginning of value`
I don't have this issue anymore with **mcr.microsoft.com/playwright** image 
-macOS Ventura 13.0 (Intel M1)

**Comment by trevesz** [comment]
It would appear so!  Issue not happening with 4.13.1 for my original containers either.  :)

**Comment by trevesz** [comment]
Unfortunately, it happened again this morning.

ERRO[0115] error waiting for container: invalid character 'c' looking for beginning of value 

All the same exact code.  Appears to still be present, though happens less frequently than before.

**Comment by dimagoltsman** [comment]
many random crashes, this version is unstable, went back to 4.12 and everything works

**Comment by riabiy-denis** [comment]
> many random crashes, this version is unstable, went back to 4.12 and everything works

Same.

**Comment by mmikhalko** [comment]
I have the same bug!

**Comment by vtsimbaluk** [comment]
Same issue, OS 13.0 (Intel)

**Comment by smailpouri** [comment]
Similar issue here with 4.13 up to 4.14. 
This issue does NOT exist on the latest 4.12.x 
Apple M1

**Comment by mmikhalko** [comment]
Issue still in 4.14 version

**Comment by helloguille** [comment]
Docker freezing since my last update, randomly.

I cannot restart it, the contenerized app (apache and php) stops working (freezes and stops responding to requests)

MacOSX Monterrey (M1)
4.13.1

**Comment by mmikhalko** [comment]
Issue still in 4.14.1 version

**Comment by helloguille** [comment]
How could I go back to 4.12.1?

**Comment by muldos** [comment]
Hello, same here apple M1 max / OS Ventura 13.0 (22A380)

**Comment by helloguille** [comment]
Downgrading to 4.12 solved the issue for me.

**Comment by yakobe** [comment]
Same issue here

**Comment by nicks** [comment]
how often do you see it happen? i ran a little script like:

```
#!/bin/bash

set -eux

docker build -t tmp .

for i in {1..50}
do
  docker run --rm tmp
done
```

with the Dockerfile and app.py at the top, and wasn't able to repro it.

**Comment by bluehz** [comment]
Unfortunately I am unable to test it as I refuse to install the updated
Docker again. It really did a number on my setup.
Lemme know when you get past this bug.

**Comment by trevesz** [comment]
@nicks it was happening on every run to me but I haven't tried it again with latest release since others keep commenting its still happening to them.  I will try it on my setup with latest build shortly and let you know.

**Comment by trevesz** [comment]
@nicks you're right...doesn't crash reliably on simple scrapes anymore...so something must have changed since the issue first showed up but it still crashes reliably for my script when handling larger pages...this sample code causes the crash reliably on my system still with 4.15.0.

Also worth noting, with all prior versions I tried, once the docker daemon has crashed, I could select Quit Docker Desktop from the menu bar icon and it would eventually exit Docker Desktop cleanly and I could restart it and run again.  With this version, trying to Quit no longer does anything and the app remains in a hung state until I reboot my entire machine.

Hope this helps track down the issue!

```
from selenium import webdriver
from selenium.webdriver.common.by import By
from time import sleep

class Scraper:
    def __init__(self, delay):
        # Let's grab our default settings for Chromium instance
        chrome_options = self.__build_chrome_options()
        self.delay = delay

        try:
            # Let's instantiate our scraper
            self.driver = webdriver.Chrome(options=chrome_options)
        except Exception as e:
            raise e

    def __build_chrome_options(self):
        chrome_options = webdriver.ChromeOptions()
        chrome_options.add_argument('--window-size=1920,1200')
        chrome_options.add_argument('--headless')
        chrome_options.add_argument('--single-process')
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-gpu')
        chrome_options.add_argument('--enable-logging')
        chrome_options.add_argument('--user-data-dir=/tmp/user1')
        chrome_options.add_argument('--v=1')
        chrome_options.add_argument('--ignore-certificate-errors')
        chrome_options.add_argument('--disable-dev-shm-usage')

        return chrome_options

    def scroll_down(self, delay):
        while True:
            # Get scroll height.
            lastHeight = self.driver.execute_script("return document.body.scrollHeight")

            # Scroll down to the bottom. 
            print(f"Last height: [{lastHeight}] - Scrolling...") # uncomment for debugging scrolling
            self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")

            # Wait to load the page
            sleep(delay/2)

            # Calculate new scroll height and compare with last scroll height
            newHeight = self.driver.execute_script("return document.body.scrollHeight")

            # If the browser hasn’t scrolled any more (i.e. it’s reached the end) then stop
            if newHeight == lastHeight:
                print("Reached bottom") # uncomment for debugging scrolling
                break

    def get_with_delay(self, request_url, delay):
        self.driver.get(request_url)
        sleep(delay)
        self.scroll_down(delay)
    
    def __del__(self):
        self.driver.quit()

# use this for testing locally
if __name__ == "__main__":
    scraper = Scraper(3)

    url = "https://www.facebook.com/ads/library/?active_status=active&ad_type=all&country=US&view_all_page_id=281735115664505&search_type=page&media_type=all"

    try:
        print("Getting url: ", url)
        scraper.get_with_delay(url, 3)
        html = scraper.driver.find_elements(By.CLASS_NAME, '_7jvw.x2izyaf.x1hq5gj4.x1d52u69')
        print(f"Found {len(html)} items.")
        for h in html:
            print(h.find_element(By.CLASS_NAME, 'xt0e3qv').text)

    except Exception as e:
        print(f'Error {e} while scraping url: {url}')
```
and just for the sake of consistency...I skipped including a simple requirements.txt file in my initial post but this is the one I'm using for the above code execution.

```
async-generator==1.10
attrs==22.1.0
certifi==2022.9.24
exceptiongroup==1.0.4
h11==0.14.0
idna==3.4
outcome==1.2.0
PySocks==1.7.1
selenium==4.6.1
sniffio==1.3.0
sortedcontainers==2.4.0
trio==0.22.0
trio-websocket==0.9.2
urllib3==1.26.13
wsproto==1.2.0
```

**Comment by Mcgoozers** [comment]
im having the same issue but im running a rocket pool ethereum node software. reverting to 4.12 is the only way i can make it not crash

**Comment by wivaku** [comment]
same behavior with `transmission-openvpn`.  No issue with 4.12. Docker crashing with 4.13 and up. Including 4.15.

**Comment by djs55** [comment]
Thanks for the reports and repro example. I have a test build which should have more reliable container TCP handling if you'd like to try it:
- [Apple Silicon](https://desktop-stage.docker.com/mac/main/arm64/93263/Docker.dmg)
- [intel](https://desktop-stage.docker.com/mac/main/amd64/93263/Docker.dmg)

Let me know if it works, and if not, please upload a diagnostic and let me know.

**Comment by djs55** [comment]
@trevesz thanks for the quick response. Looking at the logs there is more UDP activity than I expected, perhaps this is because chrome is using QUIC rather than TCP. If you have time for more experiments, perhaps try disabling QUIC and forcing TCP by adding
```
        chrome_options.add_argument('--disable-quic')
```
In the meantime I'll keep digging.

**Comment by trevesz** [comment]
@djs55 thanks for the update and suggestion.  I wasn't familiar with QUIC...great to learn about it.  And pleased to report that 2 days in a row now my script has run cleanly after adding the suggested chrome option and using your experimental build.  Will leave this open as I think this is all still partial it sounds like but standing by to test anything else if needed.

**Comment by nzcodarnoc** [comment]
I'm also experiencing this on an Intel Mac.

**Comment by bgronemanLIO** [comment]
Same issue running V.14 and V15, though on two machines we were able to resolve this by going back to version 4.12

By adding the defensive code mentioned by @trevesz [comment](https://github.com/docker/for-mac/issues/6530#issuecomment-1334404449)
And reverting back to 4.12 I was able to get around this crashing issue.

Appreciate ya'll's work on this!

**Comment by djs55** [comment]
@trevesz thanks for the update! It's good to know we're on the right track...

I've got another test build which should have both TCP and UDP working if you'd like to try it (without the QUIC workaround). Let me know what you think:

- [amd64](https://desktop-stage.docker.com/mac/main/amd64/93533/Docker.dmg)
- [arm64](https://desktop-stage.docker.com/mac/main/arm64/93533/Docker.dmg)

**Comment by bgronemanLIO** [comment]
> @trevesz thanks for the update! It's good to know we're on the right track...
> 
> I've got another test build which should have both TCP and UDP working if you'd like to try it (without the QUIC workaround). Let me know what you think:
> 
> * [amd64](https://desktop-stage.docker.com/mac/main/amd64/93533/Docker.dmg)
> * [arm64](https://desktop-stage.docker.com/mac/main/arm64/93533/Docker.dmg)

With this build and no QUIC workaround I was able to run our full test suite without Docker crashing.

@djs55 you may have found the solution here.

**Comment by wivaku** [comment]
Same here, I no longer see the issue in this build. Thanks @djs55! 

I am not using Selenium/Chrome, but transmission-openvpn (with UDP VPN). But had similar issue in v4.13 - 4.15 (and no issue in v4.12).

**Comment by trevesz** [comment]
@djs55 looks like you have a winner!  Clean run just now with your latest build above and QUIC enabled again.  :)

**Comment by dimagoltsman** [comment]
so, is this fixed now?

**Comment by Mcgoozers** [comment]
im continuing to have the same issue on 4.15... 4.12 is still stable so had to downgrade to that again

**Comment by wivaku** [comment]
> im continuing to have the same issue on 4.15... 4.12 is still stable so had to downgrade to that again. however im not running chromium im running rocketpool client

Did you try the build mentioned above: https://github.com/docker/for-mac/issues/6530#issuecomment-1344029769?

**Comment by cliss** [comment]
> Did you try the build mentioned above: https://github.com/docker/for-mac/issues/6530#issuecomment-1344029769?

I did, and it did not fix the issue.

**Comment by tabrej-alam** [comment]
Same issue with Docker.app V 4.15,4.14 and 4.13.
macOS Ventura 13.0.1, 2.4 GHz 8-Core Intel Core i9

After upgrading to Ventura, I can not even run V4.12.0. it says docker.app is damaged when I am trying to download and install 4.12.0.

**Comment by djs55** [comment]
@Mcgoozers @cliss @tabrej-alam if you're still having issues with the build mentioned here: https://github.com/docker/for-mac/issues/6530#issuecomment-1344029769 could you reproduce the upload fresh diagnostics? It is probably a different issue.

**Comment by samsawyer** [comment]
@djs55 — just a note that I tried your build ([above](https://github.com/docker/for-mac/issues/6530#issuecomment-1344029769)) and it seems to have fixed the problem I was having, and a lot of other people reporting too, at #6533 (Docker releases >4.12 crashing on MacOS under heavy network loads.)

**Comment by ylor** [comment]
> @djs55 — just a note that I tried your build ([above](https://github.com/docker/for-mac/issues/6530#issuecomment-1344029769)) and it seems to have fixed the problem I was having, and a lot of other people reporting too, at #6533 (Docker releases >4.12 crashing on MacOS under heavy network loads.)

Can confirm that this has fixed the stability issue I was experiencing that matches #6533 however I thought it was worth noting that on my machine I am seeing a 10x+ decrease in overall network throughput.

**Comment by xjose97x** [comment]
Still crashing for me every now and then. I have a M1 Pro processor.
Running Docker 4.15.0 (93002)

**Comment by kennethredler** [comment]
> @trevesz thanks for the update! It's good to know we're on the right track...
> 
> I've got another test build which should have both TCP and UDP working if you'd like to try it (without the QUIC workaround). Let me know what you think:
> 
> * [amd64](https://desktop-stage.docker.com/mac/main/amd64/93533/Docker.dmg)
> * [arm64](https://desktop-stage.docker.com/mac/main/arm64/93533/Docker.dmg)

Build is working great @djs55 ‼️

No longer seeing Docker "going out to lunch". 👍👏

When can we expect this fix in a production release? 🙏

**Comment by djs55** [comment]
@kennethredler I'm hoping to get this fix into 4.16, due in January. It's merged already but I'm on the look out for possible regressions. Speaking of which:

@ylor could you describe your setup a little more? Is it TCP or UDP? Are you using the virtualisation.framework (or Intel hyperkit or Apple Silicon qemu)? Virtualization framework with recent macOS versions should be fast. Could you share a benchmark result (e.g. with iperf?) Thanks!

**Comment by ylor** [comment]
Hey @djs55 👋, no benchmarks unfortunately -- downgraded back to 4.12. I can try and get some put together but in the meantime, my setup looks like:
- macOS 13.1
- Virtualization Framework + VirtioFS
- [Large file transfers (TCP mostly) of Linux ISOs](https://github.com/linuxserver/docker-transmission) tunneled through a [Wireguard (UDP only)](https://github.com/linuxserver/docker-wireguard) container

**Comment by kennethredler** [comment]
> @ kennethredler I'm hoping to get this fix into 4.16, due in January. It's merged already but I'm on the look out for possible regressions. Speaking of which:
> 
> @ ylor could you describe your setup a little more? Is it TCP or UDP? Are you using the virtualisation.framework (or Intel hyperkit or Apple Silicon qemu)? Virtualization framework with recent macOS versions should be fast. Could you share a benchmark result (e.g. with iperf?) Thanks!

Thanks for the info and update @djs55 . Regrettably, the real issue I'm seeing may be #6472 ("Docker is Stopping") and this build doesn't seem to resolve that. Will be looking for help over there.

**Comment by samsawyer** [comment]
@djs55, unfortunately I can confirm @ylor's regression. The set up is using docker compose to run a Wireguard VPN container (so all UDP) and funneling network traffic through that. When I do that and test using fast.com in a browser, I get 93 Mbps from fast.com under 4.12 and only 6.8 Mbps under your new 4.16 build.

To confirm, I also set up the same VPN connection using a different Wireguard container and ran iperf3 tests — here's what that looks like:

Under 4.12:
```root@2514c6071d1d/ # iperf3 -4 -V -p 5204 -c 94.154.159.137
iperf 3.12
Linux 2514c6071d1d 5.10.124-linuxkit #1 SMP Thu Jun 30 08:19:10 UTC 2022 x86_64
Control connection MSS 1368
Time: Sat, 31 Dec 2022 16:05:52 GMT
Connecting to host 94.154.159.137, port 5204
      Cookie: ujfqmc7cdnvgwhbzv6emfk5enyw3drqeyhx2
      TCP MSS: 1368 (default)
[  5] local 10.16.253.3 port 47290 connected to 94.154.159.137 port 5204
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  12.6 MBytes   105 Mbits/sec    0    334 KBytes       
[  5]   1.00-2.00   sec  12.9 MBytes   108 Mbits/sec    1    367 KBytes       
[  5]   2.00-3.00   sec  12.8 MBytes   107 Mbits/sec    0    393 KBytes       
[  5]   3.00-4.00   sec  13.7 MBytes   115 Mbits/sec    0    417 KBytes       
[  5]   4.00-5.00   sec  12.8 MBytes   107 Mbits/sec    0    438 KBytes       
[  5]   5.00-6.00   sec  10.1 MBytes  84.4 Mbits/sec    0    454 KBytes       
[  5]   6.00-7.00   sec  12.8 MBytes   107 Mbits/sec    0    474 KBytes       
[  5]   7.00-8.00   sec  12.3 MBytes   103 Mbits/sec    0    508 KBytes       
[  5]   8.00-9.00   sec  13.6 MBytes   114 Mbits/sec    0    590 KBytes       
[  5]   9.00-10.00  sec  13.6 MBytes   114 Mbits/sec    0    695 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   127 MBytes   107 Mbits/sec    1             sender
[  5]   0.00-10.00  sec   124 MBytes   104 Mbits/sec                  receiver
CPU Utilization: local/sender 1.1% (0.2%u/0.9%s), remote/receiver 3.2% (0.6%u/2.6%s)
```

and under your 4.16 build:
```root@b2cef2755e74/ # iperf3 -4 -V -p 5204 -c 94.154.159.137
iperf 3.12
Linux b2cef2755e74 5.15.49-linuxkit #1 SMP Tue Sep 13 07:51:46 UTC 2022 x86_64
Control connection MSS 1368
Time: Sat, 31 Dec 2022 15:47:33 GMT
Connecting to host 94.154.159.137, port 5204
      Cookie: a3au2ke5xuodtpy5fuajmsfu3ciebo23rdpt
      TCP MSS: 1368 (default)
[  5] local 10.15.173.84 port 49974 connected to 94.154.159.137 port 5204
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  1.52 MBytes  12.8 Mbits/sec    0    108 KBytes       
[  5]   1.00-2.00   sec  1005 KBytes  8.23 Mbits/sec   11   93.5 KBytes       
[  5]   2.00-3.00   sec   753 KBytes  6.17 Mbits/sec    7   74.8 KBytes       
[  5]   3.00-4.00   sec  1005 KBytes  8.23 Mbits/sec    0   86.8 KBytes       
[  5]   4.00-5.00   sec   753 KBytes  6.18 Mbits/sec    0   90.8 KBytes       
[  5]   5.00-6.01   sec  1005 KBytes  8.18 Mbits/sec    0   96.2 KBytes       
[  5]   6.01-7.01   sec  1.04 MBytes  8.75 Mbits/sec    5   70.8 KBytes       
[  5]   7.01-8.01   sec   753 KBytes  6.18 Mbits/sec    1   78.8 KBytes       
[  5]   8.01-9.00   sec  1005 KBytes  8.27 Mbits/sec    0   86.8 KBytes       
[  5]   9.00-10.00  sec  1.23 MBytes  10.3 Mbits/sec    0   97.5 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  9.92 MBytes  8.32 Mbits/sec   24             sender
[  5]   0.00-10.00  sec  9.56 MBytes  8.02 Mbits/sec                  receiver
CPU Utilization: local/sender 0.5% (0.1%u/0.4%s), remote/receiver 0.1% (0.1%u/0.0%s)
```

In both 4.12 and 4.16, I'm using the new Virtualization framework and running macOS Ventura 13.1.

If I can provide any additional details that would be helpful in tracking this down, happy to do what I can.

**Comment by bewareofgeek** [comment]
I tried the 4.16 test build as well on mac m1. It resolved the issue of docker going down completely when under network load, but now my total throughput from the containers is stuck at about 5MB/s. I've tested this repeatedly by having downloads in both transmission and sabnzbd. They'll each get a part of the 5MB/s and if I pause one, the other gets it's speed. I'm wondering, did you just put a limiter. Is the original issue caused by higher throughput? In any case, anxious to get a fix for this. It's sad my Pi4 is more efficient at this and it has to copy the completed downloads over the network to my media server.

**Comment by djs55** [comment]
Thanks for the info on UDP performance. I found a logging problem in that development build -- it was logging once per packet (!) so I've removed that and tidied the network logs generally. If you'd like to try a new M1 build I have one here: https://desktop-stage.docker.com/mac/main/arm64/94771/Docker.dmg This also has a fix for a bug which affected multiple concurrent TCP connections.

Could you try it and let me know what happens? For reference the logs are in `~/Library/Containers/com.docker.docker/Data/log/vm/nat.log` -- take a look to check you're not seeing excessive logging in there.

**Comment by samsawyer** [comment]
@djs55 if you have an Intel build that fixes the logging issue for UDP performance, I'd be happy to test it and report back iperf numbers again.

**Comment by bewareofgeek** [comment]
@djs55 I know I jumped in on this, but I wanted to let you know I just tried that latest build you provided. The docker crashing and bandwidth issues I was having both appear to be resolved. You're a lifesaver. I was pulling my hair out trying to figure this out. First time using docker and just needed something better than the Pi4 I was using and couldn't figure out why I was the only one seeming to have issues on such a basic setup.

**Comment by djs55** [comment]
@samsawyer I've got an Intel build here: https://desktop-stage.docker.com/mac/main/amd64/94771/Docker.dmg -- let me know what you observe!

**Comment by Neurone** [comment]
The [test build](https://github.com/docker/for-mac/issues/6530#issuecomment-1370001795) solved [this issue](https://github.com/docker/for-mac/issues/6650) on my machine!

**Comment by samsawyer** [comment]
@djs55, nice work. Just tested with the new Intel build, and here are some perf results:

```
iperf 3.12
Linux f66c6ab4be84 5.15.49-linuxkit #1 SMP Tue Sep 13 07:51:46 UTC 2022 x86_64
Control connection MSS 1368
Time: Wed, 04 Jan 2023 01:44:27 GMT
Connecting to host 94.154.159.137, port 5204
      Cookie: j2ouplljljrbdybtsmsonnooniqhbzxudj4d
      TCP MSS: 1368 (default)
[  5] local 10.6.156.58 port 53438 connected to 94.154.159.137 port 5204
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  11.8 MBytes  99.1 Mbits/sec   13    172 KBytes       
[  5]   1.00-2.00   sec  8.58 MBytes  72.0 Mbits/sec   19    124 KBytes       
[  5]   2.00-3.00   sec  9.99 MBytes  83.8 Mbits/sec    0    172 KBytes       
[  5]   3.00-4.00   sec  10.4 MBytes  87.4 Mbits/sec   11    158 KBytes       
[  5]   4.00-5.01   sec  9.87 MBytes  82.4 Mbits/sec    9    104 KBytes       
[  5]   5.01-6.00   sec  8.58 MBytes  72.4 Mbits/sec    0    152 KBytes       
[  5]   6.00-7.00   sec  11.0 MBytes  92.6 Mbits/sec    1    140 KBytes       
[  5]   7.00-8.00   sec  10.4 MBytes  87.5 Mbits/sec   19    139 KBytes       
[  5]   8.00-9.00   sec  9.93 MBytes  83.3 Mbits/sec   11    135 KBytes       
[  5]   9.00-10.00  sec  11.0 MBytes  92.6 Mbits/sec    0    186 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   102 MBytes  85.3 Mbits/sec   83             sender
[  5]   0.00-10.00  sec   100 MBytes  84.2 Mbits/sec                  receiver
CPU Utilization: local/sender 1.1% (0.2%u/0.9%s), remote/receiver 2.3% (0.4%u/1.8%s)
```

So back to near-parity with 4.12 in terms of performance; whew.

(For the sake of completeness: I also ended up trying [colima](https://github.com/abiosoft/colima) as a result of all this — and in colima, the same iperf test run with the same docker-compose.yml gives me more than double the performance, clocking in at 222 Mbps.)

But in any case, I think you've solved the speed regression from 4.12 to what's due to be merged into 4.16, and you'd already solved the crash-under-network-load issue. So thanks for tackling both of those.

**Comment by sellersj** [comment]
@djs55 thank you for your work.

4.16 is released. Does this mean all the changes are part of that release?

**Comment by djs55** [comment]
@sellersj yes, all those changes should be in 4.16. Let me know if you see anything suspicious!

**Comment by trevesz** [comment]
Awesome news.  Thank you as well for the great follow up and support @djs55!  Closing out the issue as resolved.

-tomas

**Comment by mat007** [comment]
A fix has been released in Docker Desktop 4.16. See the [release notes](https://docs.docker.com/desktop/release-notes/) for more details.

**Comment by ylor** [comment]
Thank you @samsawyer for following up with numbers and @djs55 for the hard work!
